VerilogForVerificationStyle.md

```diff
@@ -0,0 +1,863 @@
+7.  `DV_EOT_PRINT_..._CONTENTS` TLM fifo display macros
+    *   It is recommended to use these macros to display the status of the
+    various TLM fifos and queues in the scoreboard at the end of the test
+    to ensure that those data structures have been emptied. There should be no
+    pending transactions that have not yet been compared before the simulation
+    ends.
+    If used, these macros must be used during the `check_phase`.
+8.  `DV_SPINWAIT`
+    *   It is recommended to use this macro in situations that require an
+    isolation fork to disable some process after waiting for a specified
+    amount of time, as it provides a built in timer watchdog and proper process isolation:
+
+```systemverilog
+// forked threads should be labeled if they serve a specific purpose,
+// like the isolation thread below
+fork : isolation_fork
+  begin
+    fork
+      begin
+        <specified code to execute>
+      end
+      begin
+        <watchdog timer logic with error reporting>
+      end
+    join_any
+    disable fork;
+  end
+join
+```
+
+General macro use guidelines:
+
+1.  Macro naming
+    *   If the macro wraps some fundamental UVM functionality and deals
+    directly with UVM functions, like `uvm_component_new`, and so on, the
+    macro must be named all in `lower_snake_case`
+    *   If the macro expands into a block of code that does not deal directly
+    with UVM functionality, and is for convenience, like the
+    `DV_CHECK_*` macros, the macro must be named in `UPPER_SNAKE_CASE`
+2.  Macro vs. Parameters:
+    *   If the intent is to define a constant that is a basic data type or
+    that is the result of an expression involving basic data types, a parameter
+    (or localparam) must be used, depending on where the constants are being defined.
+    *   If the intent is to abstract away a code snippet or to represent a
+    slice of an array, a macro must be used.
+    If the macro is defined in a local perspective, it must be undefined at
+    the end of the file, otherwise it will pollute the global namespace.
+
+
+### Factory
+
+To use the UVM factory, these guidelines should be followed:
+
+1.  All classes must be registered with the factory. Use the
+    `uvm_object_utils` and `uvm_component_utils` macros to register all
+    non-parameterized classes, and use `uvm_object_param_utils` and
+    `uvm_component_param_utils` to register all parameterized classes.
+2.  The `create()` API must be used to create all objects and components.
+    Exceptions for this rule are any TLM related objects and covergroup objects,
+    which should be created by directly calling `new()`, as they should not be
+    registered with the factory.
+3.  When using the factory to create objects with `<type_name>::type_id::create()`,
+    the `<type_name>` must be explicitly declared, parameters representing types
+    must not be used here.
+3.  When using the factory to create objects, it is recommended to use the
```

Comment:
Evidently `.md` will heed your numbering when you pick up a new numbered list
but autoincrements for you if it is a continuation of a numbered list. So this `3`
renders as `4` (correctly), but the next `4` renders as `4` when it should be `5`.

Reply:
Ah yeah, seems I accidentally listed two items as `3.`, I've updated this list to fix the issue.